articles/service-fabric/service-fabric-reliable-services-notifications.md

```diff
@@ -17,8 +17,7 @@ Common reasons for using notifications are:
 * Building materialized views, such as secondary indexes or aggregated filtered views of the replica's state. An example is a sorted index of all keys in Reliable Dictionary.
 * Sending monitoring data, such as the number of users added in the last hour.
 
-Notifications are fired as part of applying operations. 
-Because of that, notifications should be handled as fast as possible, and synchronous events shouldn't include any expensive operations.
+Notifications are fired as a part of applying operations. On a primary replica operations are applied after quorum acknowledgment as a part of `transaction.CommitAsync()` or `this.StateManager.GetOrAddAsync()`. On secondary replicas operations are applied at replication queue data processing. Because of that, notifications should be handled as fast as possible, and synchronous events shouldn't include any expensive operations. Otherwise it could negatielly impact transaction processing time as well as replica build ups.

```

Comment:
> or `this.StateManager.GetOrAddAsync()`
> applied after quorum acknowledment

Dictionary notifications are fired only during CommitAsync().
Since naturally, all operations before CommitAsync have quorum replicated, it is true. But we don't wait for `Commit (EndTransaction)` record for quorum ack.

So, for multi Operation Tx, after all operations are replicated/logged to disk.
Today, for single operation Tx, just after logging to disk, we apply. We don't wait for quorum ack. But no one can read from memory until it is quorum acked. So, it is fine.

This is done to increase concurrency for single operation Tx -- do as much work in parallel with secondary replicas.

Reply:
@ashishnegi `this.StateManager.GetOrAddAsync()` can be called without transaction object, therefore there is no `CommitAsync()` from user perspective and notification is fired as a part of `GetOrAddAsync()` call. Isn't it?

>Today, for single operation Tx, just after logging to disk, we apply.

Does this mean that user can get notification before quorum ack for single operation transactions?

Reply:
@sachevl I thought we are only talking about dictionary notification. I think it would be best to not mention it here but later : under section: `Reliable State Manager provides notifications for the following events:` for each type.